samples/will_it_shuffle.lobster

```diff
@@ -0,0 +1,228 @@
+    0xB2ABD2,
+    0xD8DAEB,
+    0xF7F7F7,
+    0xFEE0B6,
+    0xFDB863,
+    0xE08214,
+    0xB35806,
+    0x7F3B08,
+]): color {
+    div( (_ >> 16) & 0xFF, 255),
+    div( (_ >> 8)  & 0xFF, 255),
+    div(  _        & 0xFF, 255),
+    1,
+}
+
+def spline_value(values, t):
+    // uniforem b-spline interpolation adapted from d3-scale-chromactic
+
+    // TODO can we translate from this Basis Spline (B-spline) to a cardinal
+    // spline? Or should lobster also provide a builtin for B-splines?
+
+    t = clamp(t, 0, 1)
+    let n = values.length - 1
+    let i = clamp(floor(t * n), 0, n-1)
+
+    let v1 = values[i]
+    let v2 = values[i + 1]
+    let v0 = if i > 0:     values[i - 1] else: v1 * 2 - v2
+    let v3 = if i < n - 1: values[i + 2] else: v2 * 2 - v1
+
+    let t1 = (t - div(i, n)) * n
+    let t2 = t1 * t1
+    let t3 = t2 * t1
+    return (
+        v0 * (1 - 3 * t1 + 3 * t2 - t3) +
+        v1 * (4 - 6 * t2 + 3 * t3) +
+        v2 * (1 + 3 * t1 + 3 * t2 - 3 * t3) +
+        v3 * t3
+    ) / 6
+
+def spline(t):
+    let c = spline_value(colors, t)
+    return color { // TODO wish we had vectorized clamp
+        clamp(c.red,   0, 1),
+        clamp(c.green, 0, 1),
+        clamp(c.blue,  0, 1),
+        clamp(c.alpha, 0, 1),
+    }
+
+var N = 60
+var M = 10000
+var matrix = map(N*N): 0
+var fills = map(N*N): color_black
+var color_domain_lo = log(M / N / 3)
+var color_domain_hi = log(M / N * 3)
+
+def reshuffle():
+    for(N) i: for(N) j: matrix[N*i + j] = 0
+    for(M):
+        let array = map(N): _
+        shuffle(array)
+        for(array) i, j:
+            matrix[N*i + j]++
+    for(matrix) n, k:
+        let t0 = color_domain_lo
+        let t1 = color_domain_hi
+        let v = clamp(log(n), t0, t1)
+        let t = if t0 == t1: 0.5 else: clamp((v - t0) / (t1 - t0), 0, 1)
+        fills[k] = spline(t)
+
+var shuffling = true
+
+fatal(gl_window("Will It Shuffle?", 640, 480))
+check(gl_set_font_name("data/fonts/US101/US101.TTF") and gl_set_font_size(32), "cannot load gl font")
+
+def shuffle_routine(next):
+    var last_n = 0
+    var last_m = 0
+    var last_mode = mode
+
+    forever():
+        // apply any changes in mode, N, or M
+        let changed = N != last_n or M != last_m
+        if changed:
+            let max_work = 60 * 10000
+            if N * M > max_work:
+                if N != last_n:
+                    M = floor(div(max_work, N) / 100) * 100
+                else:
+                    N = floor(div(max_work, M))
+            if N != last_n:
+                matrix = map(N*N): 0
+                fills = map(N*N): color_black
+                last_n = N
+            color_domain_lo = log(M / N / 3)
+            color_domain_hi = log(M / N * 3)
+            last_m = M
+
+        if shuffling or changed or last_mode != mode:
+            reshuffle()
+            last_mode = mode
+        next(nil)
+    return nil
+let shuffler = coroutine shuffle_routine()
+
+def viz():
+    let status = join(" ", [
```

Comment:
Doesn't this do pretty much exactly what `join` does?

https://github.com/aardappel/lobster/blob/dffd8a8944cbe5d81789b22926f725b013183cd7/dev/src/builtins.cpp#L504-L515

Reply:
Oops, had missed that one.